### ISSUE TITLE
Improve loading time of website

### ISSUE THREAD

**Issue by neuroactive** [koroonakaart/koroonakaart#233]
Results from webpagetest.org suggest that the site takes about 12 seconds to load on the first view, and about 3 seconds to load on subsequent views. This is unacceptably slow and we should take steps to improve it.

There are several reasons for the slow speed. Currently Node.js is compiling all the app code and data into two JavaScript files with a combined size of about 4.5 MB. These files include the code for the app itself, various dependencies, as well as GeoJSON data for the map and all the statistical data for the charts. Over time, the amount of data we are dealing with is going to increase, so unless we do something about it, the site is going to get progressively slower. Downloading these JavaScript files takes around 8 seconds on the first view, and the JavaScript itself takes almost 3 seconds to execute on both the first view and repeat view.

In my opinion we have reached the point where the current single-page-application structure is no longer suitable. We should be looking to move to a site architecture which cuts down the amount of front-end code we serve to the user, where we have better control over the caching of specific parts of the code, and where we can be more selective about what data we serve to the user. The performance of the site could also be improved by splitting it into separate pages, which would have benefits in terms of UX too. Rather than sending the user the data for all the charts at once, we could then send them only what they need for the current page they are on.

In the meantime, it would speed up the site if we implement Gzip compression on the JavaScript and CSS files output by Node. This would reduce the download size of the those files from around 4.8 MB to 1.8 MB, which would be a significant improvement.

https://webpagetest.org/result/210426_AiDcPB_2c863176ab6dc326d8944dd4c9939e71/

![loading_time_2021_apr_26](https://user-images.githubusercontent.com/22959023/116064109-5ebddc80-a68e-11eb-838c-025cf9fcfb94.png)

![loading_time_2021_apr_26_ii](https://user-images.githubusercontent.com/22959023/116064145-67161780-a68e-11eb-9022-041b80c25a51.png)

**Comment by dorkster100** [comment]
As i mentioned earlier, the real solution here is to package data for each chart into separate json files and lazy load them via API.

However as an intermediate solution, the numbers will get better if we simply make the data js file load asynchronously. Not that it will be a real benefit to the end user, they will get placeholders, and graphs will still take the same time to load. But the test websites will give a better result.

Even if we want to separate the page in several sub pages it still requires us to have the data file separated into packages or API that will spit out only required data. It even may make sense to actually use some database for it, redis, elastic, mongodb...

**Comment by FottyM** [comment]
> As i mentioned earlier, the real solution here is to package data for each chart into separate json files and lazy load them via API.
> 
> 
> 
> However as an intermediate solution, the numbers will get better if we simply make the data js file load asynchronously. Not that it will be a real benefit to the end user, they will get placeholders, and graphs will still take the same time to load. But the test websites will give a better result.
> 
> 
> 
> Even if we want to separate the page in several sub pages it still requires us to have the data file separated into packages or API that will spit out only required data. It even may make sense to actually use some database for it, redis, elastic, mongodb...

Agreed, I also think we could load the data asynchronously and put it in some database. Maybe put some spinners... while the user waits.

**Comment by neuroactive** [comment]
@dorkster100 I agree with you that loading the data asynchronously would be a good idea and I am going to open a new issue for it.

> However as an intermediate solution, the numbers will get better if we simply make the data js file load asynchronously. Not that it will be a real benefit to the end user, they will get placeholders, and graphs will still take the same time to load. But the test websites will give a better result.

Yes, this would be a good intermediate step, but I don't really agree with you that there would be no benefit to the user. There is a huge difference in the perceived speed of a website if the page starts rendering straight away rather than having the user look at a screen which doesn't change at all for 10 seconds. My interest is in improving the UX, not the test results per se.

> Even if we want to separate the page in several sub pages it still requires us to have the data file separated into packages or API that will spit out only required data. It even may make sense to actually use some database for it, redis, elastic, mongodb...

Right, I am going to start work on the API soon. Ultimately adding a cache may make sense, although I suppose we can manage without it for the time being.

**Comment by neuroactive** [comment]
These are the WebPageTest results after enabling Gzip. The performance metrics have improved quite a lot.

https://webpagetest.org/result/210511_AiDcBW_3cf1566bf18282ccd193bc509715ac04/

![loading_time_2021_may_11_after_gzip](https://user-images.githubusercontent.com/22959023/117825984-72fdfe00-b278-11eb-9dba-1526a823e4a7.png)

![loading_time_2021_may_11_after_gzip_ii](https://user-images.githubusercontent.com/22959023/117825991-75605800-b278-11eb-8c76-ed2d954551f5.png)

**Comment by lietu** [comment]
Basically solved by https://github.com/okestonia/koroonakaart/pull/245 without any API or database complications to worry about
